### PR TITLE
Setup update and API Token Authentication

### PR DESCRIPTION
**Pull request by gideonmaina**
## Description

The main aim of this PR is to simplify local development and implement authentication for key endpoints prone to bot crawling or misuse by clients whilst fixing a few issues related to dependencies and migrations. The affected endpoints/views in API `v1` are `data`, `now` and `node`.  This will attempt to reduce the lag observed when making calls to the API.
 
Fixes issues #134, #135, #136

## Setup changes
- The `feinstaub sensors app` was removed from the working tree as pip install takes care of that.
- The migration file `*04` in sensorsafrica app was removed as it is applied before a dependency

Tests for these changes were both done using local and staging databases.  

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Screenshots

Screenshot 2024-08-16 at 11 08 02

Screenshot 2024-08-16 at 16 04 36

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have successfully build the the project and performed necessary tests

### CODE REVIEW

**Review by DavidTheProgrammer — APPROVED** [review]
Looks good

File: Dockerfile [hunk 1]
```diff
@@ -13,6 +13,9 @@ VOLUME [ "/src/logs" ]
 # Copy the current directory contents into the container at sensorsafrica
 ADD . /src/
 
+# Upgrade pip from trusted hosts
+RUN python -m pip install --trusted-host pypi.python.org --trusted-host files.pythonhosted.org --trusted-host pypi.org --upgrade pip
+
 # Upgrade pip and setuptools
 RUN pip install -q -U pip setuptools
```
Comment (kilemensi) [inline Dockerfile:20]: Why 2 ways of upgrading `pip`?

File: README.md [hunk 2]
```diff
@@ -26,8 +26,12 @@ CREATE DATABASE sensorsafrica;
 CREATE USER sensorsafrica WITH ENCRYPTED PASSWORD 'sensorsafrica';
 GRANT ALL PRIVILEGES ON DATABASE sensorsafrica TO sensorsafrica;
 ```
-
 - Migrate the database; `python manage.py migrate`
+- Create super user for admin login; `python manage.py createsuperuser` 
+  username: `sensorsafrica`
+  email: blank
+  password: `sensorsafrica`
```
Comment (kilemensi) [inline README.md:33]: Must username/password be `sensorsafrica`?

File: docker-compose.yml [hunk 3]
```diff
@@ -2,15 +2,20 @@ version: '3.3'
 
 services:
   rabbitmq:
-    image: rabbitmq:3.5.1
+    image: rabbitmq:3.12.7-management
```
Comment (kilemensi) [inline docker-compose.yml:5]: Is 3.12 the version used in PROD as well?

File: docker-compose.yml [hunk 4]
```diff
@@ -21,15 +26,15 @@ services:
...
     environment:
-      SENSORSAFRICA_DATABASE_URL: postgres://sensorsafrica:sensorsafrica@postgres:5432/sensorsafrica
-      SENSORSAFRICA_READ_DATABASE_URLS: postgres://sensorsafrica:sensorsafrica@postgres:5432/sensorsafrica
-      SENSORSAFRICA_RABBITMQ_URL: amqp://sensorsafrica:sensorsafrica@rabbitmq//
-      SENSORSAFRICA_FLOWER_ADMIN_USERNAME: admin
-      SENSORSAFRICA_FLOWER_ADMIN_PASSWORD: password
+      SENSORSAFRICA_DATABASE_URL: ${SENSORSAFRICA_DATABASE_URL}
+      SENSORSAFRICA_RABBITMQ_URL: ${SENSORSAFRICA_RABBITMQ_URL}
+      SENSORSAFRICA_FLOWER_ADMIN_USERNAME: ${SENSORSAFRICA_FLOWER_ADMIN_USERNAME}
+      SENSORSAFRICA_FLOWER_ADMIN_PASSWORD: ${SENSORSAFRICA_FLOWER_ADMIN_PASSWORD}
```
Comment (kilemensi) [inline docker-compose.yml:37]: Is there a default for any of these?

File: sensorsafrica/settings.py [hunk 5]
```diff
@@ -113,13 +113,6 @@
 
 DATABASES = {"default": dj_database_url.parse(DATABASE_URL), }
 
-READ_DATABASE_URLS = os.getenv("SENSORSAFRICA_READ_DATABASE_URLS", DATABASE_URL).split(",")
-
-for index, read_database_url in enumerate(READ_DATABASE_URLS,start=1):
-    DATABASES[f"read_replica_{index}"] = dj_database_url.parse(read_database_url)
-
-DATABASE_ROUTERS = ["sensorsafrica.router.ReplicaRouter", ]
```
Comment (kilemensi) [inline sensorsafrica/settings.py:121]: 👍🏽

**Review by kilemensi — COMMENTED** [review]
👍🏽 

---

If I understand what we're trying to do, then I see this as breaking change i.e. the new secure endpoints are no longer v1 but v2 (or whatever v we're currently on).

My approach would be:
1. Create new v2 endpoints that work exactly like the equivalent v1 endpoints **but** secure.
2. Test and deploy backend to PROD.
3. Upgrade the front-end up to point to v2 endpoints.
4. Test and deploy front-end to PROD.
5. Remove or redirect unsecure v1 endpoints.
6. Test and deploy backend to PROD.
7. Profit

File: docker-compose.yml (hunk 4 above)
Comment (gideonmaina) [reply docker-compose.yml:37]: Was previously. 
```
  SENSORSAFRICA_DATABASE_URL: postgres://sensorsafrica:sensorsafrica@postgres:5432/sensorsafrica
  SENSORSAFRICA_RABBITMQ_URL: amqp://sensorsafrica:sensorsafrica@rabbitmq:5672
  SENSORSAFRICA_FLOWER_ADMIN_USERNAME: admin
  SENSORSAFRICA_FLOWER_ADMIN_PASSWORD: password```

Comment (gideonmaina) [reply docker-compose.yml:37]: This will be updated in the README for local setup.

**Comment by thepsalmist** [comment]
> 👍🏽
> 
> If I understand what we're trying to do, then I see this as breaking change i.e. the new secure endpoints are no longer v1 but v2 (or whatever v we're currently on).
> 
> My approach would be:
> 
> 1. Create new v2 endpoints that work exactly like the equivalent v1 endpoints **but** secure.
> 2. Test and deploy backend to PROD.
> 3. Upgrade the front-end up to point to v2 endpoints.
> 4. Test and deploy front-end to PROD.
> 5. Remove or redirect unsecure v1 endpoints.
> 6. Test and deploy backend to PROD.
> 7. Profit

Cool, makes sense @gideonmaina I'll setup a review to plan for this implementation approach

**Comment by kilemensi** [comment]
PS: Not sure what `SENSORSAFRICA_CELERY_SLACK_WEBHOOK_FAILURES_ONLY` is for. If we need to send errors notifications to Slack, the best approach it to integrate with Sentry and Sentry will send notifications over email, Slack, etc.

File: Dockerfile (hunk 1 above)
Comment (gideonmaina) [reply Dockerfile:20]: 👍🏽 the upgrade flag will be removed to just install the setup tools

File: README.md (hunk 2 above)
Comment (gideonmaina) [reply README.md:33]: Not really. Let's say it a convention. Any thoughts?

**Comment by thepsalmist** [comment]
Changes should be merged in #138

**Pull request closed without merging**